src/odemis/gui/comp/overlay/rectangle.py

```diff
@@ -0,0 +1,356 @@
+
+    def clear(self):
+        """Remove all rectangles and update canvas."""
+        for rectangle in self._rectangles:
+            rectangle.active.value = False
+        self._rectangles.clear()
+
+    def remove_overlay(self, overlay: RectangleOverlay):
+        """Remove a rectangle's overlay and update canvas."""
+        if overlay in self._rectangles:
+            overlay.active.value = False
+            self._rectangles.remove(overlay)
+            self.cnvs.remove_world_overlay(overlay)
+            self.cnvs.request_drawing_update()
+
+    def on_enter(self, evt):
+        if self.active.value:
+            self.cnvs.set_default_cursor(wx.CURSOR_CROSS)
+        else:
+            WorldOverlay.on_enter(self, evt)
+
+    def on_leave(self, evt):
+        if self.active.value:
+            self.cnvs.reset_default_cursor()
+        else:
+            WorldOverlay.on_leave(self, evt)
+
+    def _lock_rectangles(self, flag=True):
+        """Lock or unlock rectangles to process on_left_down and on_left_up events."""
+        for rectangle in self._rectangles:
+            rectangle.locked.value = flag
+
+    def _activate_rectangles(self, flag=True):
+        """Activate or de-activate the rectangles."""
+        for rectangle in self._rectangles:
+            rectangle.active.value = flag
+
+    def _on_tool(self, selected_tool):
+        """Update the overlay when it's active and tools change."""
+        if selected_tool == guimodel.TOOL_RECTANGLE:
+            self.active.value = True
+            self._lock_rectangles(False)
+            # Make the last created rectangle active
+            if self._rectangles:
+                last_rectangle = self._rectangles[-1]
+                last_rectangle.active.value = True
+        elif selected_tool == guimodel.TOOL_NONE:
+            self.active.value = False
+            self._lock_rectangles(True)
+            self._activate_rectangles(False)
+            self.cnvs.reset_default_cursor()
+
+    def _get_rectangle(self, evt) -> Union[RectangleOverlay, None]:
+        """
+        Find a rectangle corresponding to the given on_left_down event position.
+        Returns: (RectangleOverlay or None): the most appropriate rectangle.
+            If no rectangle is found, it returns None.
+        """
+        if self._rectangles:
+            pos = self.cnvs.view_to_phys(evt.Position, self.cnvs.get_half_buffer_size())
+            for rectangle in self._rectangles[::-1]:
+                if util.is_point_in_rect(pos, rectangle.coordinates.value):
+                    return rectangle
+        return None
+
+    def on_left_down(self, evt):
+        """Start drawing a rectangle if the overlay is active and there is no selected rectangle."""
+        if not self.active.value:
+            return super().on_left_down(evt)
+
+        self._selected_rectangle = self._get_rectangle(evt)
+        if self._selected_rectangle is None:
+            rectangle = RectangleOverlay(self.cnvs)
```

Comment:
As discussed, I think that you can mostly get it working by using Overlays. However it can be tricky to properly handle hover (with a correct icon). It might also make things quite complicated if you want to introduce groups.
So I'd recommend having a separate class, not directly an Overlay, and having this RectangesOverlay class call the draw of each instance of Rectangle/Shape.